File: internal/ui/echo_filter.go
package ui

import (
	"strings"

	"github.com/charmbracelet/log"
)

// echoFilter strips the shell's local echo of a just-submitted command from the
// raw PTY stream so only the host command block (pushBlock) shows the text.
//
// Windows PowerShell syntax-highlights the echo with CSI color codes even under
// -NoProfile. Matching is a streaming state machine: skip CSI while reading the
// plaintext command, then drop through the trailing newline. Leading CSI
// (cursor hide, etc.) is suppressed while armed; non-matching plaintext before
// the command is passed through without disarming (avoids giving up on noise).
type echoFilter struct {
	want  []rune // plaintext command to suppress
	pos   int    // matched rune count into want
	phase int    // 0 = match command, 1 = drop through newline
	armed bool
	seen  int // raw bytes consumed while armed (give-up budget)

	// CSI / ESC parser while suppressing
	escState int // 0 none, 1 saw ESC, 2 CSI params, 3 OSC
}

// Max PTY bytes to examine while armed before giving up (don't strand the stream).
const echoFilterGiveUp = 4096

const (
	echoPhaseMatch = 0
	echoPhaseNL    = 1
)

// arm prepares to suppress echo of cmd (may be multi-line). Empty cmd disarms.
func (f *echoFilter) arm(cmd string) {
	*f = echoFilter{} // full reset
	if stringsTrimSpace(cmd) == "" {
		return
	}
	// Normalize to a single logical echo unit. Multi-line: match first line
	// only for now (PS continuation prompts are a separate problem); remaining
	// lines still show once — better than double-printing the primary line.
	norm := strings.ReplaceAll(cmd, "\r\n", "\n")
	norm = strings.ReplaceAll(norm, "\r", "\n")
	if i := strings.IndexByte(norm, '\n'); i >= 0 {
		norm = norm[:i]
	}
	norm = strings.TrimRight(norm, " \t")
	if norm == "" {
		return
	}
	f.want = []rune(norm)
	f.armed = true
	f.phase = echoPhaseMatch
	log.Debug("echo filter armed", "cmd", norm)
}

// feed consumes PTY bytes and returns bytes that should reach the VT parser.
func (f *echoFilter) feed(in []byte) []byte {
	if !f.armed || len(f.want) == 0 {
		return in
	}
	if len(in) == 0 {
		return nil
	}

	var out []byte
	for i := 0; i < len(in); i++ {
		b := in[i]

		// Once finished, pass the remainder through.
		if !f.armed {
			out = append(out, in[i:]...)
			break
		}

		f.seen++
		if f.seen > echoFilterGiveUp {
			log.Debug("echo filter give up", "seen", f.seen)
			out = append(out, in[i:]...)
			f.disarm()
			break
		}

		// ESC / CSI / OSC handling while armed:
		//
		// PowerShell wraps local-echo in SGR color CSI — suppress those once we
		// have started matching the command text (pos>0) or are draining the
		// trailing newline (phase NL).
		//
		// BUT: when pos==0 (not yet matching), pass CSI through. Otherwise a
		// Unix `clear` / terminfo wipe that arrives *before* the echoed command
		// (or when echo is off) is swallowed and the VT never clears.
		if f.escState != 0 || b == 0x1b {
			passThrough := f.pos == 0 && f.phase == echoPhaseMatch
			if passThrough {
				out = append(out, b)
				_ = f.consumeEsc(b) // keep parser in sync; bytes already emitted
				continue
			}
			if f.consumeEsc(b) {
				// still inside escape — drop (PS color / cursor hide around echo)
				continue
			}
			// escape finished — drop the final byte too
			continue
		}

		switch f.phase {
		case echoPhaseMatch:
			// CR/LF: while not yet matching, pass through (real prior output).
			// Once matching, CR is end-of-echo noise; LF resets a partial match.
			if b == '\r' {
				if f.pos == 0 {
					out = append(out, b)
				}
				continue
			}
			if b == '\n' {
				if f.pos == 0 {
					out = append(out, b)
					continue
				}
				// Mid-command newline shouldn't happen for single-line; reset.
				f.pos = 0
				continue
			}
			// Other controls: pass when idle, drop/reset when mid-match.
			if b < 0x20 {
				if f.pos == 0 {
					out = append(out, b)
					continue
				}
				f.pos = 0
				continue
			}

			want := f.want[f.pos]
			// Compare as rune; input is UTF-8 — for ASCII commands byte==rune.
			r, n := decodeRuneByte(in, i)
			if n > 1 {
				// Multi-byte rune in stream
				if r == want {
					f.pos++
					i += n - 1
					if f.pos >= len(f.want) {
						f.phase = echoPhaseNL
						f.pos = 0
					}
					continue // suppressed
				}
				// Mismatch: if we haven't started, pass this rune through.
				if f.pos == 0 {
					out = append(out, in[i:i+n]...)
					i += n - 1
					continue
				}
				// Started matching then diverged — pass pending? We already
				// suppressed matched bytes (unrecoverable). Reset and pass current.
				log.Debug("echo filter reset mid-match", "got", string(r), "want", string(want))
				f.pos = 0
				out = append(out, in[i:i+n]...)
				i += n - 1
				continue
			}

			// Single-byte
			if rune(b) == want {
				f.pos++
				if f.pos >= len(f.want) {
					f.phase = echoPhaseNL
					f.pos = 0
				}
				continue // suppress
			}
			if f.pos == 0 {
				// Leading plaintext that isn't our command (real output early).
				out = append(out, b)
				continue
			}
			// Divergence after partial match.
			log.Debug("echo filter reset mid-match", "got", string(b), "want", string(want))
			f.pos = 0
			out = append(out, b)

		case echoPhaseNL:
			// Drop everything until we consume a newline (end of echo line).
			if b == '\n' {
				f.disarm()
				log.Debug("echo filter matched and suppressed")
				continue // drop the newline too
			}
			// Drop CR, trailing CSI already handled above, drop leftover junk
			// on the echo line (should be nothing for PS).
			continue
		}
	}
	return out
}

func (f *echoFilter) disarm() {
	f.armed = false
	f.want = nil
	f.pos = 0
	f.phase = 0
	f.escState = 0
	f.seen = 0
}

// status reports suppressor state for the MCP diag bridge.
func (f *echoFilter) status() (armed bool, cmd string, phase int) {
	if !f.armed {
		return false, "", 0
	}
	return true, string(f.want), f.phase
}

// consumeEsc advances ESC/CSI/OSC state. Returns true if still inside a sequence.
func (f *echoFilter) consumeEsc(b byte) bool {
	switch f.escState {
	case 0:
		if b == 0x1b {
			f.escState = 1
			return true
		}
		return false
	case 1: // after ESC
		switch b {
		case '[':
			f.escState = 2
			return true
		case ']':
			f.escState = 3
			return true
		default:
			// ESC X single-byte (or charset) — done after this byte
			f.escState = 0
			return false
		}
	case 2: // CSI ... final byte 0x40-0x7e
		if b >= 0x40 && b <= 0x7e {
			f.escState = 0
			return false
		}
		return true
	case 3: // OSC ... BEL or ST
		if b == 0x07 {
			f.escState = 0
			return false
		}
		// ST is ESC\ — if we see ESC, go to state 1 and let next iter handle;
		// but ST end is ESC \ : handle ESC inside OSC by staying, next byte \\.
		if b == 0x1b {
			f.escState = 4
			return true
		}
		return true
	case 4: // OSC almost ST: expect '\'
		f.escState = 0
		return false
	default:
		f.escState = 0
		return false
	}
}

// decodeRuneByte decodes one UTF-8 rune starting at i, or one byte on error.
func decodeRuneByte(in []byte, i int) (rune, int) {
	if i >= len(in) {
		return 0, 0
	}
	b := in[i]
	if b < 0x80 {
		return rune(b), 1
	}
	// Minimal UTF-8 decode (2-3 byte); enough for command lines we care about.
	if b&0xe0 == 0xc0 && i+1 < len(in) {
		return rune(b&0x1f)<<6 | rune(in[i+1]&0x3f), 2
	}
	if b&0xf0 == 0xe0 && i+2 < len(in) {
		return rune(b&0x0f)<<12 | rune(in[i+1]&0x3f)<<6 | rune(in[i+2]&0x3f), 3
	}
	return rune(b), 1
}

File: chrome/src/echo_filter.rs
//! Streaming echo suppressor — product port of `internal/ui/echo_filter.go`.
//!
//! After a warp-bar submit, the host injects a command block and arms this
//! filter so the shell's local echo of the same line does not double-print.

/// Max PTY bytes examined while armed before giving up.
pub const ECHO_FILTER_GIVE_UP: usize = 4096;

const PHASE_MATCH: i32 = 0;
const PHASE_NL: i32 = 1;

/// Strips local echo of a just-submitted command from the raw PTY stream.
#[derive(Clone, Debug, Default)]
pub struct EchoFilter {
    want: Vec<char>,
    pos: usize,
    phase: i32,
    armed: bool,
    seen: usize,
    /// 0 none, 1 saw ESC, 2 CSI params, 3 OSC, 4 OSC ST almost done
    esc_state: i32,
}

impl EchoFilter {
    pub fn new() -> Self {
        Self::default()
    }

    /// Prepare to suppress echo of `cmd`. Empty / whitespace disarms.
    pub fn arm(&mut self, cmd: &str) {
        *self = Self::default();
        let mut norm = cmd.replace("\r\n", "\n").replace('\r', "\n");
        if let Some(i) = norm.find('\n') {
            norm.truncate(i);
        }
        let norm = norm.trim_end_matches([' ', '\t']);
        if norm.is_empty() {
            return;
        }
        self.want = norm.chars().collect();
        self.armed = true;
        self.phase = PHASE_MATCH;
    }

    pub fn disarm(&mut self) {
        *self = Self::default();
    }

    /// `(armed, cmd, phase)` for MCP diag.
    pub fn status(&self) -> (bool, String, i32) {
        if !self.armed {
            return (false, String::new(), 0);
        }
        (true, self.want.iter().collect(), self.phase)
    }

    pub fn is_armed(&self) -> bool {
        self.armed
    }

    /// Consume PTY bytes; return bytes that should reach the VT decoder.
    pub fn feed(&mut self, input: &[u8]) -> Vec<u8> {
        if !self.armed || self.want.is_empty() {
            return input.to_vec();
        }
        if input.is_empty() {
            return Vec::new();
        }

        let mut out = Vec::with_capacity(input.len());
        let mut i = 0;
        while i < input.len() {
            if !self.armed {
                out.extend_from_slice(&input[i..]);
                break;
            }

            let b = input[i];
            self.seen += 1;
            if self.seen > ECHO_FILTER_GIVE_UP {
                out.extend_from_slice(&input[i..]);
                self.disarm();
                break;
            }

            if self.esc_state != 0 || b == 0x1b {
                let pass_through = self.pos == 0 && self.phase == PHASE_MATCH;
                if pass_through {
                    out.push(b);
                    let _ = self.consume_esc(b);
                    i += 1;
                    continue;
                }
                if self.consume_esc(b) {
                    i += 1;
                    continue;
                }
                // escape finished — drop final byte too
                i += 1;
                continue;
            }

            match self.phase {
                PHASE_MATCH => {
                    if b == b'\r' {
                        if self.pos == 0 {
                            out.push(b);
                        }
                        i += 1;
                        continue;
                    }
                    if b == b'\n' {
                        if self.pos == 0 {
                            out.push(b);
                            i += 1;
                            continue;
                        }
                        self.pos = 0;
                        i += 1;
                        continue;
                    }
                    if b < 0x20 {
                        if self.pos == 0 {
                            out.push(b);
                            i += 1;
                            continue;
                        }
                        self.pos = 0;
                        i += 1;
                        continue;
                    }

                    let want = self.want[self.pos];
                    let (r, n) = decode_rune_byte(input, i);
                    if n > 1 {
                        if r == want {
                            self.pos += 1;
                            i += n;
                            if self.pos >= self.want.len() {
                                self.phase = PHASE_NL;
                                self.pos = 0;
                            }
                            continue;
                        }
                        if self.pos == 0 {
                            out.extend_from_slice(&input[i..i + n]);
                            i += n;
                            continue;
                        }
                        self.pos = 0;
                        out.extend_from_slice(&input[i..i + n]);
                        i += n;
                        continue;
                    }

                    if char::from(b) == want {
                        self.pos += 1;
                        i += 1;
                        if self.pos >= self.want.len() {
                            self.phase = PHASE_NL;
                            self.pos = 0;
                        }
                        continue;
                    }
                    if self.pos == 0 {
                        out.push(b);
                        i += 1;
                        continue;
                    }
                    self.pos = 0;
                    out.push(b);
                    i += 1;
                }
                PHASE_NL => {
                    if b == b'\n' {
                        self.disarm();
                        i += 1;
                        continue;
                    }
                    i += 1;
                }
                _ => {
                    out.push(b);
                    i += 1;
                }
            }
        }
        out
    }

    /// Returns true if still inside an escape sequence.
    fn consume_esc(&mut self, b: u8) -> bool {
        match self.esc_state {
            0 => {
                if b == 0x1b {
                    self.esc_state = 1;
                    return true;
                }
                false
            }
            1 => match b {
                b'[' => {
                    self.esc_state = 2;
                    true
                }
                b']' => {
                    self.esc_state = 3;
                    true
                }
                _ => {
                    self.esc_state = 0;
                    false
                }
            },
            2 => {
                if (0x40..=0x7e).contains(&b) {
                    self.esc_state = 0;
                    false
                } else {
                    true
                }
            }
            3 => {
                if b == 0x07 {
                    self.esc_state = 0;
                    false
                } else if b == 0x1b {
                    self.esc_state = 4;
                    true
                } else {
                    true
                }
            }
            4 => {
                self.esc_state = 0;
                false
            }
            _ => {
                self.esc_state = 0;
                false
            }
        }
    }
}

fn decode_rune_byte(input: &[u8], i: usize) -> (char, usize) {
    if i >= input.len() {
        return ('\0', 0);
    }
    let b = input[i];
    if b < 0x80 {
        return (char::from(b), 1);
    }
    if b & 0xe0 == 0xc0 && i + 1 < input.len() {
        let cp = (u32::from(b & 0x1f) << 6) | u32::from(input[i + 1] & 0x3f);
        return (char::from_u32(cp).unwrap_or('\u{FFFD}'), 2);
    }
    if b & 0xf0 == 0xe0 && i + 2 < input.len() {
        let cp = (u32::from(b & 0x0f) << 12)
            | (u32::from(input[i + 1] & 0x3f) << 6)
            | u32::from(input[i + 2] & 0x3f);
        return (char::from_u32(cp).unwrap_or('\u{FFFD}'), 3);
    }
    (char::from(b), 1)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn suppresses_highlighted_ps_echo() {
        let raw = b"\x1b[?25l\x1b[93mWrite-Output \x1b[37mhello\r\n\x1b[?25h\x1b[mhello\r\n ";
        let mut f = EchoFilter::new();
        f.arm("Write-Output hello");
        let got = f.feed(raw);
        let s = String::from_utf8_lossy(&got);
        assert!(!s.contains("Write-Output"), "command leak: {s:?}");
        assert!(s.contains("hello"), "output missing: {s:?}");
        assert!(!f.is_armed());
    }

    #[test]
    fn chunked() {
        let mut f = EchoFilter::new();
        f.arm("echo hi");
        let parts: &[&[u8]] = &[
            b"\x1b[?25l",
            b"\x1b[93mecho \x1b[37mhi\r",
            b"\n\x1b[?25h\x1b[mhi\r\n ",
        ];
        let mut got = Vec::new();
        for p in parts {
            got.extend(f.feed(p));
        }
        let s = String::from_utf8_lossy(&got);
        assert!(!s.contains("echo"), "command leak: {s:?}");
        assert!(s.contains("hi"), "output missing: {s:?}");
    }

    #[test]
    fn leading_output_passes() {
        let mut f = EchoFilter::new();
        f.arm("Get-Date");
        let raw = b"NOTICE\r\n\x1b[93mGet-Date\r\n\x1b[mthedate\r\n";
        let got = f.feed(raw);
        let s = String::from_utf8_lossy(&got);
        assert!(s.contains("NOTICE"), "leading lost: {s:?}");
        assert!(!s.contains("Get-Date"), "command leak: {s:?}");
        assert!(s.contains("thedate"), "output missing: {s:?}");
    }

    #[test]
    fn empty_arm() {
        let mut f = EchoFilter::new();
        f.arm("   ");
        let raw = b"hello\r\n";
        assert_eq!(f.feed(raw), raw);
    }

    #[test]
    fn passes_clear_csi_before_echo() {
        let mut f = EchoFilter::new();
        f.arm("clear");
        let raw = b"\x1b[H\x1b[2J\x1b[3Jclear\r\n";
        let got = f.feed(raw);
        let s = String::from_utf8_lossy(&got);
        assert!(s.contains("\x1b[2J"), "clear CSI swallowed: {s:?}");
        assert!(!s.contains("clear"), "command leak: {s:?}");
    }

    #[test]
    fn status_reports_armed() {
        let mut f = EchoFilter::new();
        assert_eq!(f.status(), (false, String::new(), 0));
        f.arm("ls -la");
        let (armed, cmd, phase) = f.status();
        assert!(armed);
        assert_eq!(cmd, "ls -la");
        assert_eq!(phase, 0);
    }
}
